File: eshop/client_nextjs/src/lib/storefront/categories.ts
import type { Category } from "@/types/storefront";

export const ALL_CATEGORY: Category = {
  id: 0,
  name: "All",
  slug: "all",
  description: "All products",
  icon: "Grid2X2",
  image_url: null,
  display_order: 0,
  is_featured: true,
  parent_id: null,
};

export function withAllCategoryFirst(categories: Category[]) {
  return [ALL_CATEGORY, ...categories.filter((category) => category.slug !== ALL_CATEGORY.slug)];
}

export function rootCategories(categories: Category[]) {
  return withAllCategoryFirst(categories).filter(
    (category) => category.slug === ALL_CATEGORY.slug || category.parent_id === null,
  );
}

export function categoryChildren(categories: Category[], parentId: number) {
  return categories
    .filter((category) => category.parent_id === parentId && category.slug !== ALL_CATEGORY.slug)
    .sort((a, b) => a.display_order - b.display_order || a.name.localeCompare(b.name));
}

export function hasCategoryChildren(categories: Category[], categoryId: number) {
  return categories.some((category) => category.parent_id === categoryId);
}

export function categoryDescendantIds(categories: Category[], categoryId: number) {
  const descendants = new Set<number>();
  const pending = [categoryId];
  while (pending.length) {
    const current = pending.pop();
    if (current === undefined) continue;
    for (const child of categoryChildren(categories, current)) {
      if (descendants.has(child.id)) continue;
      descendants.add(child.id);
      pending.push(child.id);
    }
  }
  return descendants;
}

export function categoryDepth(categories: Category[], category: Category) {
  const byId = new Map(categories.map((item) => [item.id, item]));
  const visited = new Set<number>();
  let depth = 0;
  let current = category;
  while (current.parent_id !== null && !visited.has(current.id)) {
    visited.add(current.id);
    const parent = byId.get(current.parent_id);
    if (!parent) break;
    depth += 1;
    current = parent;
  }
  return depth;
}

File: eshop/backend_django/api/services/categories.py
from collections import defaultdict, deque

from api.models import ProductCategory
from api.models import Product


def active_category_subtree_ids(category):
    """Return an active category and every active descendant in one category query."""
    if not category.is_active:
        return []

    children_by_parent = defaultdict(list)
    for category_id, parent_id in ProductCategory.objects.filter(
        is_active=True
    ).values_list("id", "parent_id"):
        children_by_parent[parent_id].append(category_id)

    subtree_ids = []
    pending = deque([category.id])
    visited = set()
    while pending:
        category_id = pending.popleft()
        if category_id in visited:
            continue
        visited.add(category_id)
        subtree_ids.append(category_id)
        pending.extend(children_by_parent.get(category_id, ()))
    return subtree_ids


def filter_products_by_category_subtree(queryset, category):
    return queryset.filter(category_id__in=active_category_subtree_ids(category))


def store_category_hierarchy(trader):
    """Relevant active leaf categories plus active ancestors, excluding broken inactive paths."""
    categories = list(ProductCategory.objects.all().only(
        "id", "parent_id", "name", "slug", "description", "icon", "image",
        "display_order", "is_featured", "is_active",
    ))
    by_id = {category.id: category for category in categories}
    direct_ids = set(Product.objects.filter(
        trader=trader, status=Product.Status.ACTIVE, category__isnull=False
    ).values_list("category_id", flat=True))
    included = set()
    for category_id in direct_ids:
        path = []
        current = by_id.get(category_id)
        valid = True
        visited = set()
        while current:
            if not current.is_active or current.id in visited:
                valid = False
                break
            visited.add(current.id)
            path.append(current.id)
            current = by_id.get(current.parent_id)
        if valid:
            included.update(path)
    return sorted(
        (by_id[category_id] for category_id in included),
        key=lambda category: (category.display_order, category.name, category.id),
    )
